### ISSUE TITLE
Add another comparison, but with offline learning

### ISSUE THREAD

**Issue by Expertium** [open-spaced-repetition/fsrs-vs-sm15#22]
I understand that this version uses online learning (parameters are updated after each repetition) to make things fair. However, I think it would be interesting to see the full potential of FSRS, as currently this would give the users the wrong impression that FSRS is only as good as SM-15, when in reality it's more accurate.

**Comment by Expertium** [comment]
Unrelated, but also add MAE to the comparison.

**Comment by L-M-Sherlock** [comment]
I plan to add "experience replay" to the online learning version. It allow FSRS online to reuse the learnt data.

For more details about "experience replay", please see: [Experience Replay for Continual Learning](https://arxiv.org/abs/1811.11682)

**Comment by L-M-Sherlock** [comment]
The result of offline learning could only show the potential capacity to fit data. If we need to evaluate the performance of FSRS as real as possible, we should employ [time series split](https://scikit-learn.org/stable/modules/cross_validation.html#time-series-split) at least.

**Comment by Expertium** [comment]
But offline learning is how FSRS is actually optimized in practice, isn't it?

**Comment by L-M-Sherlock** [comment]
> But offline learning is how FSRS is actually optimized in practice, isn't it?

If we add replay to online learning, it will approach the actual optimization in FSRS Optimizer. 

In FSRS Optimizer, we train FSRS with the entire dataset five times. Then it will be used to predict the memory in practice.

In "experience replay", we could train FSRS online with the learnt data multiple times, too.

**Comment by Expertium** [comment]
Ok, maybe I wasn't clear. I want to see two comparisons. One is as fair as possible, and the other shows the performance of FSRS that matches the actual, real performance.
So one comparison should have online learning without experience replay (because SuperMemo doesn't use it), and the other should have offline learning exactly the way it is currently implemented in the real optimizer.
The first comparison is for the sake of fairness, and the second comparison is for the sake of demonstrating the real performance of FSRS as is.

**Comment by L-M-Sherlock** [comment]
> Unrelated, but also add MAE to the comparison.

MAE is biased in our case.

<img width="913" alt="image" src="https://github.com/open-spaced-repetition/fsrs-vs-sm15/assets/32575846/515f7edc-5e3d-4fd1-a5ca-d88e21961e2b">

For more details, please see: https://www.semanticscholar.org/paper/Metrics-for-Evaluation-of-Student-Models-Pel%C3%A1nek/71c98355e21c0b30a9e3d545508afa6b53f1e974

**Comment by L-M-Sherlock** [comment]
> learning without experience replay (because SuperMemo doesn't use it)

It is possible that SuperMemo has used it, because SuperMemo stored many data in its matrixes. The online version of FSRS is just used to avoid data leaking - FSRS shouldn't know the data from the future.

> A scenario where a model is trained to predict stock prices using historical data. If the training data happens to contain information on future stock prices, the model will not be able to generalize to new data and will give overly optimistic performance metrics during testing.
> Source: https://medium.com/@chabavictor7/data-leakage-in-machine-learning-d2ae0b3cd6ca

**Comment by Expertium** [comment]
> MAE is biased in our case.

I just want it so I can make a reddit post, and since I used MAE last time, I have to use it again, for consistency. And in my post I want to include both online-FSRS and offline-FSRS. Online for a fair comparison, offline to demonstrate the actual accuracy that real FSRS provides.

**Comment by L-M-Sherlock** [comment]
OK, I will restore the offline version and add MAE to evaluation tonight.

**Comment by L-M-Sherlock** [comment]
Experience Replay improved FSRS Online significantly:
- #24

**Comment by Expertium** [comment]
I'm still not sure if it's a good idea. I doubt SuperMemo uses it, so it will make the comparison unfair.

**Comment by L-M-Sherlock** [comment]
The fairness is only related to information. FSRS still has the same information as the SuperMemo. Experience Replay is just a method to deal with the known information. It doesn't leak the data from the future.

**Comment by Expertium** [comment]
Yes, but ideally we want to keep the training method similar as well.

**Comment by L-M-Sherlock** [comment]
Actually, Wozniak didn't reveal the training method of SM-15. Do you read the updated part in FSRS Online? I can explain the experience replay to you.

**Comment by Expertium** [comment]
> Actually, Wozniak didn't reveal the training method of SM-15.

I don't have the links right now, but I believe he finds global parameters by doing offline optimization with many collections, and then those parameters are incrementally updated on each user's collection.
